### PR TITLE
feat: supervised reduce method

### PR DESCRIPTION
**Pull request by db091756**
### PR Type

- Feature
- Tests

### Description

Adds solver.

### How Has This Been Tested?

Additional unit tests pass.

### Does this PR introduce a breaking change?

(Write your answer here.)

### Screenshots

(Write your answer here.)

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: coreax/solvers/coresubset.py [hunk 1]
```diff
@@ -458,3 +486,301 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+    """
+    # Extract all the possible "next" coreset arrays
+    extract_indices = (candidate_coresets[:, :, None], candidate_coresets[:, None, :])
+    coreset_feature_gramians = feature_gramian[extract_indices]
+    coreset_response_gramians = response_gramian[extract_indices]
+    coreset_cmes = training_cme[extract_indices]
+
+    # Invert the coreset feature gramians
+    if inverse_approximator is None:
+        inverse_approximator = LeastSquareApproximator(random_key)
```
Comment (pc532627) [inline coreax/solvers/coresubset.py:529]: This line is not hit by a unit test

File: coreax/solvers/coresubset.py [hunk 2]
```diff
@@ -458,3 +486,301 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+            training_cme = jnp.pad(training_cme, [(0, 1)], mode="constant")
+
+        else:
+            feature_gramian = solver_state.feature_gramian
+            response_gramian = solver_state.response_gramian
+            training_cme = solver_state.training_cme
+
+        # Sample the indices to be considered at each iteration ahead of time.
+        if self.batch_size is not None and self.batch_size < num_data_pairs:
+            batch_size = self.batch_size
```
Comment (pc532627) [inline coreax/solvers/coresubset.py:691]: This line is not hit by a unit test

File: coreax/solvers/coresubset.py [hunk 3]
```diff
@@ -458,3 +486,301 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+            random_key=self.random_key,
+            max_index=num_data_pairs,
+            batch_size=batch_size,
+            num_batches=self.coreset_size,
+        )
+
+        # Insert coreset indices into the batch indices to avoid degrading the CMMD
+        # (only has an effect when refining)
+        if self.batch_size is not None:
+            batch_indices = batch_indices.at[:, -1].set(coreset_indices)
```
Comment (pc532627) [inline coreax/solvers/coresubset.py:704]: This line is not hit by a unit test

File: coreax/solvers/coresubset.py [hunk 4]
```diff
@@ -458,3 +486,301 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+                response_gramian,
+                training_cme,
+                self.regularisation_parameter,
+                updated_identity,
+                self.inverse_approximator,
+            )
+
+            # Find the optimal replacement coreset index, ensuring we don't pick an
+            # already chosen point if we want the indices to be unique.
+            if self.unique:
```
Comment (pc532627) [inline coreax/solvers/coresubset.py:742]: I think all unit tests have unique as True - if I am correct it might be worth testing this case where unique is false

**Review by pc532627 — CHANGES_REQUESTED** [review]
Generally looks good, just a few documentation points and test coverage points

File: coreax/solvers/coresubset.py [hunk 5]
```diff
@@ -61,12 +71,30 @@ class RPCholeskyState(eqx.Module):
     gramian_diagonal: Array
 
 
+class GreedyCMMDState(eqx.Module):
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/solvers/coresubset.py:74]: Yeah the name is quite bad currently, maybe CMMDHerding makes more sense

File: coreax/solvers/coresubset.py [hunk 6]
```diff
@@ -458,3 +486,301 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+        axis2=2,
+    )
+    term_3s = jnp.trace(
+        coreset_cmes @ inverse_coreset_feature_gramians, axis1=1, axis2=2
+    )
+    return term_2s - 2 * term_3s
+
+
+# pylint: disable=too-many-locals
+class GreedyCMMD(
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/solvers/coresubset.py:552]: No, it isn't as the various operations e.g. the inverse, are affected by the presence of additional data pairs, and weights are not taken into account here, so zero-weighting doesn't solve the issue.

File: coreax/solvers/coresubset.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/solvers/coresubset.py:74]: Changed name to `ConditionalKernelHerding`

File: examples/greedy_cmmd_approximate_inverse.py
[the remark this answers is not in the document]
Comment (db091756) [reply examples/greedy_cmmd_approximate_inverse.py:1]: I've added the basic example as it seems more digestible given that CMMD as a concept it already a lot without the approximate inverse stuff.

File: coreax/solvers/coresubset.py [hunk 7]
```diff
@@ -458,3 +486,301 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+    with uniform weights.
+
+    :param random_key: Key for random number generation
+    :param feature_kernel: :class:`~coreax.kernel.Kernel` instance implementing a kernel
+        function :math:`k: \mathbb{R}^d \times \mathbb{R}^d \rightarrow \mathbb{R}`
+        on the feature space
+    :param response_kernel: :class:`~coreax.kernel.Kernel` instance implementing a
+        kernel function :math:`k: \mathbb{R}^p \times \mathbb{R}^p \rightarrow
+        \mathbb{R}` on the response space
+    :param num_feature_dimensions: An integer representing the dimensionality of the
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/solvers/coresubset.py:592]: Old code! Good spot

File: coreax/solvers/coresubset.py [hunk 8]
```diff
@@ -458,3 +486,301 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+        if self.inverse_approximator is None:
+            self.inverse_approximator = LeastSquareApproximator(self.random_key)
+
+    @override
+    def reduce(
+        self,
+        dataset: _SupervisedData,
+        solver_state: Union[GreedyCMMDState, None] = None,
+    ) -> tuple[Coresubset[_SupervisedData], GreedyCMMDState]:
+        initial_coresubset = _initial_coresubset(-1, self.coreset_size, dataset)
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/solvers/coresubset.py:621]: We fill with -1's as that index corresponds to grabbing the zero values from the padding in lines 690-692.

File: coreax/solvers/coresubset.py (hunk 1 above)
Comment (db091756) [reply coreax/solvers/coresubset.py:529]: Removed default value as this is not meant to be used by users and also got rid of the `__post_init__` as pyrght complains about it constantly and it is easy to get around it

File: tests/integration/test_david_map_reduce_weighted.py
[the remark this answers is not in the document]
Comment (db091756) [reply tests/integration/test_david_map_reduce_weighted.py:1]: I've added this but the error comes from the `import cv2` line which is ran even if the test is skipped

**Review by pc532627 — DISMISSED** [review]

File: documentation/source/conf.py [hunk 9]
```diff
@@ -169,6 +169,8 @@
     ("py:class", "jaxtyping.Shaped[Array, 'n']"),
     ("py:class", "jaxtyping.Shaped[ndarray, 'n']"),
     ("py:class", "jax._src.typing.SupportsDType"),
+    ("py:class", "numpy.bool"),
```
Comment (tp832944) [inline documentation/source/conf.py:172]: I don't think this line is required. I've added the next one as part of #675. Maybe safest to delete both of these from this PR to pre-empty junk and/or merge conflicts.

**Comment by db091756** [comment]
Not to be merged in
